library/std/src/sys/windows/stdio.rs

```diff
@@ -254,15 +324,21 @@ fn utf16_to_utf8(utf16: &[u16], utf8: &mut [u8]) -> io::Result<usize> {
     Ok(written)
 }
 
+impl IncompleteUtf8 {
+    pub const fn new() -> IncompleteUtf8 {
+        IncompleteUtf8 { bytes: [0; 4], len: 0 }
+    }
+}
+
 impl Stdout {
     pub const fn new() -> Stdout {
-        Stdout
+        Stdout { incomplete_utf8: IncompleteUtf8::new() }
     }
 }
 
 impl io::Write for Stdout {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        write(c::STD_OUTPUT_HANDLE, buf)
+        write(c::STD_OUTPUT_HANDLE, buf, &mut self.incomplete_utf8)
```

Comment:
An incomplete write will leave data in buffer even after `flush`. What behaviour would be appropriate for flush?

Reply:
This works similar to how the console output works on Macos and Linux, only that there the terminal itself does the buffering. E.g. this correctly prints "hello worldö" on the terminal on Linux and Macos:
```rust
use std::io::{self, Write};

fn main() -> io::Result<()> {
    io::stdout().write_all(b"hello world\xc3")?;
    io::stdout().flush()?;
    io::stdout().write_all(b"\xb6")?;
    Ok(())
}
```

Reply:
The difference is that once you flush there's no data buffered in the program anymore.

In general, the semantics of buffering incomplete UTF-8 seem reasonable to me. I think we'll need to talk about how this interacts with flush, though; I'm hesitant to hold on to buffered data of any kind after a flush.

Possible semantics:
1) Hold onto the incomplete character even after a flush.
2) Emit the invalid UTF-8 error when flushing an incomplete character.
3) Write a replacement character or similar.

I don't want to do (3); that seems wrong. We shouldn't hide errors.

Either (1) or (2) seems *valid*, but which one is more *useful* behavior? I suspect we want (1).